GoMeet/src/main/webapp/js/vote-meeting-time.js

```diff
@@ -0,0 +1,174 @@
+let MAX_VOTES = 3; // Users can vote up to MAX_VOTES times.
+
+/**
+ * Function called on page load of meeting-event.html
+ * to fetch all relevant MeetingTime data from servlets 
+ * and generate voting form for the MeetingTimes.
+ */
+async function displayMeetingTimeForm() {
+  let meetingEventId = getMeetingEventId();
+  let meetingEvent = await MeetingEventDAO.fetchMeetingEvent(meetingEventId);
+  let meetingTimeIds = meetingEvent['meetingTimeIds']; 
+
+  if (('status' in meetingEvent && parseInt(meetingEvent.status) !== 200) 
+      || meetingTimeIds === null) {
+    // Execution cannot continue unless meetingTimeIds can be retrieved. 
+    // Must terminate script with an Error. 
+    throw new Error(generateErrorMessage(meetingEvent) + " - MeetingEventId: " + meetingEventId);
+  }; 
+  
+  // Prepare for re-render: reset the 'meeting-times-table' table 
+  // to JUST the table headers.
+  let table = document.getElementById('meeting-times-table');
+  for (let i = table.rows.length - 1; i > 0; i--) {
+    table.deleteRow(i);
+  }
+  await fetchAndProcess(meetingTimeIds);
+}
+
+/**
+ * Returns the logged in user's email from the LoginServlet, or
+ * null if the user is not logged in.
+ * @returns The email of the current logged in user.
+ */
+async function getLoggedInUser() {
+  let status = await LoginStatus.doGet('/user-status');
+  if (status.loggedIn === 'true') {
+    return status.userEmail;
+  } else {
+    return null;
+  }
+}
+
+/**
+ * Fetches the data from the MeetingTime entities represented 
+ * by the meetingTimeIds, and pre-processes the data.
+ * @param {Array[String]} meetingTimeIds The array of meetingTimeIds 
+ * that are the meeting times associated with this Meeting
+ * @returns The array of timeData retrieved via GET request from the 
+ * '/meeting-time' endpoint, in the format: 
+ * [{meetingTimeId: id, datetime: datetimeStr, voteCount: count, voters: [voters]}]
+ */
+async function fetchAndProcess(meetingTimeIds) {
+  let currentUser = await getLoggedInUser();
+  if (currentUser === null) {
+    return null; // Do not generate the voting form if the user is not logged in.
+  }
+
+  if (!(meetingTimeIds instanceof Array)) {
+    throw new Error(INVALID_PARAM_TYPE);
+  }
+
+  let timeData = [];
+  let votedTimes = new Set(); // Times that the user has voted for, so cannot vote them again.
+  for (let i = 0; i < meetingTimeIds.length; i++) {
+    let time = await MeetingTimeDAO.fetchMeetingTime(meetingTimeIds[i]);
+    if ('status' in time && parseInt(time.status) !== 200) {
+      // NOTE: Error responses from the DAO are internal errors that the user cannot deal with, 
+      // so don't alert them, but also don't crash the program or stop execution either.
+      // Hence not throwing an error (which stops execution), but logging it.
+      console.error(generateErrorMessage(time) + " - MeetingTimeId: " + meetingTimeIds[i]);
```

Comment:
Optional: I feel like if there is a problem with the server then it would be better to alert users. For example, similar to messages when you visit websites that say something like, "Servers down, please try again later."

Reply:
The main reason I use `console.error` here instead of throwing an error and alerting the users is that I don't want to stop execution (I'm handling the case where there could be some meetingTimeIds in the array that are invalid, and others that are valid. I want to display all the valid ones, so I don't want to stop execution on any one invalid ID).

I also considered not throwing an error, but simply put an error message as part of the UI (like you mention) for times that couldn't be fetched. But I was thinking for UI neatness sake, it may be better not have the message there, because it comes back to the fact that, even if the user knew that one of the meetingTimes could not be fetched, they wouldn't be able to fix the problem. 
(poor user 😢)

Reply:
Yup, looks like you've thought about it well. Thanks for the explanation (poor user indeed 😢).